9651 scalar current date

BaurzhanSakhariev opened the pull request:
## Summary of the changes / Why this improves CrateDB

- Added date type
- Added curdate scalar function returning current date in UTC

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

crate-jenkins commented:
Can one of the admins verify this patch?

infoverload commented on docs/general/builtins/scalar.rst:
```diff
@@ -938,6 +938,20 @@ fractional seconds (milliseconds) are given.
+``CURRENT_DATE``
+----------------
+
+The ``CURRENT_DATE`` expression returns the date in UTC timezone at the time the SQL statement was handled. Clock
```
Thanks for your contribution!

Would you mind wrapping this to 79 chars per line?

mfussenegger commented on server/src/main/java/io/crate/types/DateType.java:
```diff
@@ -0,0 +1,120 @@
+import java.time.LocalDate;
+import java.time.ZoneOffset;
+
+public class DateType extends DataType<LocalDate>
+    implements FixedWidthType, Streamer<LocalDate> {
```
You'll also need to add a corresponding `PGType` implementation so that values of type `DateType` can be sent via the PostgreSQL wire protocol.

See https://github.com/crate/crate/blob/0772f2d9b3d76bb8b0ad66de96257151d4ebf2c4/server/src/main/java/io/crate/protocols/postgres/types/PGTypes.java#L42

https://github.com/crate/crate/blob/1b7b2bfdd4e950068be7f11e69c48585369d0767/server/src/main/java/io/crate/protocols/postgres/types/PGType.java#L33

If we stick to using `LocalDate` as the internal type for `date` then we'll also need an implementation for how the data type is seralized to HTTP clients. See https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/types/DataTypeXContentExtension.java

I wonder if we should instead stick to `Long` as internal type. That would reduce the amount of allocations on internal streaming operations.  Currently `readValueFrom` and `writeValueTo` is quite heavy.

BaurzhanSakhariev replied:
Could you please elaborate about having Long as internal type? I guess it means I need to replace all LocalDate usages to simple Long read/writes/casts, revert fixed size estimation to Long.size, but does it mean I will not have to change DataTypeXContentExtension.java anymore?

I will address tests & PG types & ddl doc comments accordingly to the outcome of that comment

mfussenegger commented on server/src/main/java/io/crate/types/DateType.java:
```diff
@@ -0,0 +1,120 @@
+
+    public static final int ID = 23;
+    public static final DateType INSTANCE = new DateType();
+    public static final int DATE_SIZE = (int) RamUsageEstimator.shallowSizeOfInstance(Long.class); // Value is written/read long number representing epoch milliseconds.
```
I think `LocalDate` is bigger than just a `Long` - this would have to be adjusted if we stick with `LocalDate`

BaurzhanSakhariev replied:
Upd - As you suggested, I'm using now Long as internal date, reverted DATE_SIZE to Long size

mfussenegger commented on server/src/main/java/io/crate/types/DateType.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    @Override
+    public LocalDate implicitCast(Object value) throws IllegalArgumentException, ClassCastException {
```
Could you add some tests for the cases that are handled here?

BaurzhanSakhariev replied:
Done

mfussenegger commented on server/src/main/java/io/crate/types/DataTypes.java:
```diff
@@ -76,6 +77,8 @@
     public static final TimestampType TIMESTAMPZ = TimestampType.INSTANCE_WITH_TZ;
     public static final TimestampType TIMESTAMP = TimestampType.INSTANCE_WITHOUT_TZ;
 
+    public static final DateType DATE = DateType.INSTANCE;
```
Could you also add the type to `STORAGE_UNSUPPORTED` for now and add documentation to https://github.com/crate/crate/blob/1b7b2bfdd4e950068be7f11e69c48585369d0767/docs/general/ddl/data-types.rst#L36-L48 ?

BaurzhanSakhariev replied:
Done

BaurzhanSakhariev replied:
@mfussenegger, I added DATE type to the DataTypes.STORAGE_UNSUPPORTED and updated DataTypeTesting ALL_TYPES_EXCEPT_ARRAYS -  `ALL_TYPES_EXCEPT_ARRAYS.removeIf(DataTypes.STORAGE_UNSUPPORTED :: contains).`

Basically, on my machine `LuceneQueryBuilderIntegrationTest testNullOperators` was failing with INTERVAL - I noticed, that it's also UNSUPPORTED so commented it out. `removeIf `covers INTERVAL exclusion but I decided keep `ALL_TYPES_EXCEPT_ARRAYS.add(DataTypes.INTERVAL)` just in case and commented it out, so that it's not silently excluded.

INTERVAL is being selected constantly by randomness.getRandom - I don't know is it my local issue or some extra sees is needed but anyway it's not failing anymore.

mfussenegger reviewed (commented):
Thank you for your contribution.

There are a couple of parts that are missing to add a new type, so I've left a couple of comments. 

We'd probably also need to follow up and support the date time in some artihmetic functions. E.g. to support something like `current_date + '1 day'::interval`. But I think that could be done in a follow up PR if you're interested.

mechanomi commented on docs/appendices/release-notes/unreleased.rst:
```diff
@@ -53,6 +53,8 @@ None
 Changes
 =======
 
+- Added ``current_date`` scalar function.
```
```suggestion
- Added ``CURRENT_DATE`` scalar function.
```

upper case for consistency

mechanomi reviewed (commented):
thanks for the PR!

I want to add a quick note to follow up on @mfussenegger's comment. last week, I added a TIP to the `CURRENT_TIMESTAMP` subsection about `CURRENT_TIMESTAMP - '1 day'::interval` (see https://github.com/crate/crate/commit/de7cd71d37d14128702e98d61ebc84b95a893acd)

as the `CURRENT_DATE` subsection is being added directly below this, it might be a good idea to add a short note saying either that `CURRENT_DATE - '1 day'::interval` is supported or is not supported. might be overkill though?

BaurzhanSakhariev reviewed (commented):
> thanks for the PR!
> 
> I want to add a quick note to follow up on @mfussenegger's comment. last week, I added a TIP to the `CURRENT_TIMESTAMP` subsection about `CURRENT_TIMESTAMP - '1 day'::interval` (see [de7cd71](https://github.com/crate/crate/commit/de7cd71d37d14128702e98d61ebc84b95a893acd))
> 
> as the `CURRENT_DATE` subsection is being added directly below this, it might be a good idea to add a short note saying either that `CURRENT_DATE - '1 day'::interval` is supported or is not supported. might be overkill though?

Thanks for the comments! Regarding upper case - done, but I used actual name CURDATE - I added big comment below with all my concern and one of them is that CURRENT_DATE name was failing tests while CURDATE passed them. 

TIPS - I think everything related to +/- intreval has to be done in a separate PR in order to keep the current small.


I would like to highlight potential issues with the PR. I think it's fine for now but worth mentioning anyway.

1. Initially I used "current_date" name for the new scalar function. My test were failing and it took a while before I tried to use different name and same code worked with name "curdate".

As far as I understand, "current_time" is reserved word (saw it in SqlBaseParser) and probably having a new function with that name causes parsing failure. 

Also, I saw  builder.append("current_date"); in ExpressionFormatter, the only git commit for that line shows "add the presto sql parser".

While usage of curdate passes test, I found it confusing that current_date is a new function but I cannot use that name.  

2. I haven't updated method visitCurrentTime in ExpressionAnalyzer class.It has switch (node.getType())  and that type enum already has DATE - looks like it was a placeholder for the future and I do have to add 
  case DATE:
                    funcName = CurrentTimeFunction.NAME;
                    break; but I decided to clarify that.

3. DataTypes class has 2 maps:
TYPE_IDS_TO_MAPPINGS for mapping type to ES type and I mapped new type to "date" - same as existing types TIMESTAMPZ and TIMESTAMP. 

However, there is also a reversed map MAPPING_NAMES_TO_TYPES - where ES type is mapped to Crate type. 
We cannot have 2 "date" keys and that's why I didn't an entry here. I think mapping ES date to Crate TIMESTAMPZ makes sense as TIMESTAMPZ is the most specific among DATE, TIMESTAMP, TIMESTAMPZ types. 

I believe it's fine but I have to explicitly highlight it since it looks like a place where new type might leave a trace but it was not done.

4. New types' ID field value - I spent some time trying to figure out a pattern or registry of values as I had a feeling that value must be unique throughout all types but the only approach was looking over all type files and checking their ids and keeping some unused number. There are some maps with IDS but one cannot use them as a quick ID-dictionary since they are given not as numbers but in Type.ID form. I hope my approach is fine and I haven't missed some file or even script for assigning those ids.

seut commented:
@BaurzhanSakhariev 

> I haven't updated method visitCurrentTime in ExpressionAnalyzer class.It has switch (node.getType()) and that type enum already has DATE - looks like it was a placeholder for the future and I do have to add
case DATE:
funcName = CurrentTimeFunction.NAME;
break; but I decided to clarify that.

Yes we'd have to add here a new case for `DATE` which must then result in the related `current_date` function name.



>    DataTypes class has 2 maps:
    TYPE_IDS_TO_MAPPINGS for mapping type to ES type and I mapped new type to "date" - same as existing types TIMESTAMPZ and TIMESTAMP.
>
> However, there is also a reversed map MAPPING_NAMES_TO_TYPES - where ES type is mapped to Crate type.
We cannot have 2 "date" keys and that's why I didn't an entry here. I think mapping ES date to Crate TIMESTAMPZ makes sense as TIMESTAMPZ is the most specific among DATE, TIMESTAMP, TIMESTAMPZ types.

Yes good spot. As the underlaying field mappers (of ES) do not make any difference between `DATE`  and `TIMESTAMP`, we can not simply map SQL `DATE` to a storage field type. Because of that I suggest to not support storing SQL `DATE` column types for now, we'd need to think more about a working solution here, if any.
To prevent that, the entry here https://github.com/crate/crate/pull/10978/files#diff-63dbc4faae4044b73e5c10f6ecf640d41a16284a9615d40e0c59e2ef1517219eR445 must be removed.

> New types' ID field value - I spent some time trying to figure out a pattern or registry of values as I had a feeling that value must be unique throughout all types but the only approach was looking over all type files and checking their ids and keeping some unused number. There are some maps with IDS but one cannot use them as a quick ID-dictionary since they are given not as numbers but in Type.ID form. I hope my approach is fine and I haven't missed some file or even script for assigning those ids.

Yes there is no real patter, it should just not be used already. So your chosen ID looks fine.

BaurzhanSakhariev commented:
Thanks for the comment, added switched case and removed map entry.

seut commented on server/src/main/java/io/crate/protocols/postgres/types/PGTypes.java:
```diff
@@ -56,6 +56,7 @@
         .put(DataTypes.TIMETZ, TimeTZType.INSTANCE)
         .put(DataTypes.TIMESTAMPZ, TimestampZType.INSTANCE)
         .put(DataTypes.TIMESTAMP, TimestampType.INSTANCE)
+        .put(DataTypes.DATE, TimestampType.INSTANCE)
```
To support the `DATE` type on the postgres protocol, a dedicate PgType must be implemented using the same OID postgres is using (`1082`).

BaurzhanSakhariev replied:
Done

seut commented on server/src/main/java/io/crate/protocols/postgres/types/PGTypes.java:
```diff
@@ -72,6 +73,7 @@
         .put(new ArrayType<>(DataTypes.BOOLEAN), PGArray.BOOL_ARRAY)
         .put(new ArrayType<>(DataTypes.TIMESTAMPZ), PGArray.TIMESTAMPZ_ARRAY)
         .put(new ArrayType<>(DataTypes.TIMESTAMP), PGArray.TIMESTAMP_ARRAY)
+        .put(new ArrayType<>(DataTypes.DATE), PGArray.TIMESTAMP_ARRAY)
```
Same here. must be a dedicated date array instance.

BaurzhanSakhariev replied:
Done. OID is 1182 for pg date array.

seut commented on server/src/main/resources/sql_features.tsv:
```diff
@@ -176,6 +176,7 @@ F051	Basic date and time	5	Explicit CAST between datetime types and character st
 F051	Basic date and time	6	CURRENT_DATE	NO		current_timestamp
 F051	Basic date and time	7	LOCALTIME	NO		
 F051	Basic date and time	8	LOCALTIMESTAMP	NO		
+F051	Basic date and time	9	DATE data type and CURDATE scalar function	NO		
```
Does this sub-feature exists? PostgreSQL only lists 8 subfeatures for `F051`. 
But what about subfeature `1` and `6`?

BaurzhanSakhariev replied:
Sorry, I misunderstood the pattern here - was checking other commits and thought it's kind of registry of added functions. Will address that - basically, it's about changing 1 and 6 from NO to YES but before I had to double check that I did another things correctly.

BaurzhanSakhariev replied:
Removed line with that extra feature but haven't changed 1 and 6 features to YES - not sure whether there is an extra step besides that PR to count them activated.

seut replied:
It's fine (and also preferable) to do that with the commit which is implementing the features (so this PR).

BaurzhanSakhariev replied:
Done

seut commented on server/src/test/java/io/crate/protocols/postgres/types/DateTypeTest.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Test
+    public void testDecodeUTF8TextWithUnexpectedFormat() {
+        expectedException.expect(DateTimeParseException.class);
```
Please use `assertThrows` instead as using the rule is deprecated.

BaurzhanSakhariev replied:
done

seut commented on server/src/test/java/io/crate/testing/DataTypeTesting.java:
```diff
@@ -27,22 +27,7 @@
-import io.crate.types.ShortType;
-import io.crate.types.StringType;
-import io.crate.types.TimestampType;
+import io.crate.types.*;
```
We prefer explicit imports over wildcards, please revert.

BaurzhanSakhariev replied:
Done.

seut commented on server/src/test/java/io/crate/types/DateTypeTest.java:
```diff
@@ -0,0 +1,54 @@
+        ClassCastException thrown = assertThrows(ClassCastException.class,
+            () -> DateType.INSTANCE.implicitCast("not-a-number"));
+
+        assertTrue(thrown.getMessage().equals("Can't cast 'not-a-number' to date"));
```
You could use `io.crate.testing.Asserts.assertThrows()` instead.

BaurzhanSakhariev replied:
Done

seut commented on server/src/test/java/io/crate/types/DateTypeTest.java:
```diff
@@ -0,0 +1,54 @@
+public class DateTypeTest {
+
+    @Test
+    public void test() {
```
```suggestion
    public void test_cast_from_invalid_string() {
```

BaurzhanSakhariev replied:
I renamed to camelCase style in order to keep it consistent with my other tests. I saw both underscore and camelCase test names and thought both are acceptable. Let me know if underscore is preferred I will update all my test names.

seut replied:
We have no strict rule, we just tend to the underscore format for some time now as we think its more readable specially with long test names which we tend to do to express the what-to-test in a better way. But its also fine to use CamelCase here.

BaurzhanSakhariev replied:
I will use underscore from now on)

seut commented on server/src/main/java/io/crate/types/DateType.java:
```diff
@@ -0,0 +1,121 @@
+            return null;
+        } else if (value instanceof java.util.Date) {
+            return ((java.util.Date) value).getTime();
+        }  else if (value instanceof java.sql.Date) {
```
The value can never be of type `java.sql.Date` or `java.util.Date` so these types must not be handled.

BaurzhanSakhariev replied:
Done

seut commented on server/src/test/java/io/crate/types/DateTypeTest.java:
```diff
@@ -0,0 +1,54 @@
+import static org.hamcrest.Matchers.nullValue;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.junit.jupiter.api.Assertions.*;
```
Also here, please use explicit imports.

BaurzhanSakhariev replied:
Done

seut reviewed (commented):
Thanks for your efforts. I've added another round of suggestions/comments.

BaurzhanSakhariev commented on docs/general/builtins/scalar.rst:
```diff
@@ -938,6 +938,21 @@ fractional seconds (milliseconds) are given.
 
+.. _current_date:
+
+``CURDATE``
```
Done

seut replied:
```suggestion
``CURDATE()``
```

BaurzhanSakhariev commented on server/src/main/java/io/crate/protocols/postgres/types/DateType.java:
```diff
@@ -0,0 +1,79 @@
+    Object decodeUTF8Text(byte[] bytes) {
+        String s = new String(bytes, StandardCharsets.UTF_8);
+
+        LocalDateTime dt = LocalDateTime.parse(s, ISO_DATE_PARSER);
```
Done. We don't need second formatter indeed.

BaurzhanSakhariev commented on server/src/main/java/io/crate/types/DateType.java:
```diff
@@ -0,0 +1,121 @@
+            return ((java.sql.Date) value).getTime();
+        } else if (value instanceof String) {
+            try {
+                return Long.parseLong((String) value);
```
Done. Added test as well.

BaurzhanSakhariev commented on server/src/test/java/io/crate/expression/scalar/CurrentDateFunctionTest.java:
```diff
@@ -0,0 +1,38 @@
+    public void testCallsWithinStatementAreIdempotent() {
+        assertEvaluate("CURDATE() = CURDATE()", true);
+    }
+
```
Done

BaurzhanSakhariev commented on server/src/test/java/io/crate/protocols/postgres/types/DateTypeTest.java:
```diff
@@ -0,0 +1,47 @@
+        expectedException.expect(DateTimeParseException.class);
+        DateType.INSTANCE.decodeUTF8Text("2016.06.28".getBytes(UTF_8));
+    }
+
```
Done. Thanks for the comment. actually caught a bug with a new test case.

seut commented on docs/general/builtins/scalar.rst:
```diff
@@ -938,6 +938,34 @@ fractional seconds (milliseconds) are given.
     depends on the system clock and the JVM implementation.
 
 
+.. _current_date:
```
```suggestion
.. _curdate:
```

seut commented on docs/general/builtins/scalar.rst:
```diff
@@ -938,6 +938,34 @@ fractional seconds (milliseconds) are given.
+synopsis::
+
+    CURDATE()
+
```
```suggestion

.. _current_date:

```

seut commented on docs/appendices/release-notes/unreleased.rst:
```diff
@@ -53,7 +53,7 @@ None
 =======
 
-None
+- Added ``CURDATE`` :ref:`scalar` function.
```
```suggestion
- Added :ref:`curdate` and :ref:`current_date` :ref:`scalar` functions.
```

seut commented on docs/general/builtins/scalar.rst:
```diff
@@ -938,6 +938,34 @@ fractional seconds (milliseconds) are given.
+``CURDATE``
+----------------
+
+The ``CURDATE`` expression returns the date in UTC timezone
+at the time the SQL statement was handled.
+Clock time is looked up at most once within the scope of a single query, to
+ensure that multiple occurrences of ``CURRENT_DATE`` evaluate to the
+same value.
```
The ``CURDATE()`` scalar function is an alias of the :ref:`current_date` expression. See :ref:`current_date`.

seut commented on server/src/main/resources/sql_features.tsv:
```diff
@@ -168,12 +168,12 @@ F041	Basic joined table	5	Outer joins can be nested	YES
 F051	Basic date and time	4	Comparison predicate on DATE, TIME, and TIMESTAMP data types	YES		
 F051	Basic date and time	5	Explicit CAST between datetime types and character string types	YES		
-F051	Basic date and time	6	CURRENT_DATE	NO		current_timestamp
+F051	Basic date and time	6	CURRENT_DATE	YES		current_timestamp
```
```suggestion
F051	Basic date and time	6	CURRENT_DATE	YES
```

seut reviewed (commented):
Looks really good already, I've added final documentation suggestions.

seut commented:
retest this please

seut commented:
@BaurzhanSakhariev Thanks a lot for addressing all comments.
Unfortunately the `:server:checkstyleMain` task complains now and fails, also the `itest` task fails because of the newly added `Date` and `Array(Date)` PG types, see line `163`, in `docs/interfaces/postgres.rst`.
Could you have a look on these? Thanks!

btw. some `hdfs` related test also fails cause it uses an old lib version which isn't online anymore. We've fixed this already on master, so rebasing your branch with the latest master will should solve that.

BaurzhanSakhariev commented:
I'm checking locally, some tests are failing with timeout (looks like I need to configure environment) for example `InternalTestCluster.assertAfterTest` but couldn't find non-environmental failure locally.

Thanks for pointing out the specific file (I couldn't see it inGitHub report as I need to log in to Jenkins) I will take a look for that specific case.

BaurzhanSakhariev commented:
I fixed `server:checkstyleMain` and updated `postgres.rst` with 2 entries for the new date and corresponding array type.

However, neither itest nor app:run are not working on my machine. I updated `postgres.rst` by manually filling all columns and ordering by OID.

My OS is 64-bit Windows 10 and problem with `itest `is that `blackbox/bootstrap.sh` uses `venv/bin/pip` whereas for Windows `bin `should be replaced by `Scripts`. 

I tried that and also changed` commandLine "$projectDir/bin/test-docs"` to `commandLine "sh", "$projectDir/bin/test-docs"` but still faced an issue with "Failed to import test module test_docs" - probably `test-docs` script also has to be adjusted for Windows.

Regarding `app:run`
<details>
  <summary>it fails with </summary>
  
  ```
Exception in thread "main" java.lang.ExceptionInInitializerError
	at java.base/jdk.internal.icu.text.NormalizerBase$NFKDMode.getNormalizer2(NormalizerBase.java:305)
	at java.base/jdk.internal.icu.text.NormalizerBase.normalize(NormalizerBase.java:457)
	at java.base/jdk.internal.icu.text.NormalizerBase.normalize(NormalizerBase.java:461)
	at java.base/java.text.Normalizer.normalize(Normalizer.java:159)
	at java.base/sun.security.x509.AVA.toRFC2253CanonicalString(AVA.java:987)
	at java.base/sun.security.x509.RDN.toRFC2253StringInternal(RDN.java:437)
	at java.base/sun.security.x509.RDN.toRFC2253String(RDN.java:417)
	at java.base/sun.security.x509.X500Name.getRFC2253CanonicalName(X500Name.java:724)
	at java.base/sun.security.x509.X500Name.equals(X500Name.java:416)
	at java.base/sun.security.pkcs.PKCS7.getCertificate(PKCS7.java:694)
	at java.base/sun.security.pkcs.SignerInfo.getCertificate(SignerInfo.java:253)
	at java.base/sun.security.pkcs.SignerInfo.verify(SignerInfo.java:401)
	at java.base/sun.security.pkcs.PKCS7.verify(PKCS7.java:578)
	at java.base/sun.security.pkcs.PKCS7.verify(PKCS7.java:595)
	at java.base/sun.security.pkcs.SignerInfo.getTimestamp(SignerInfo.java:566)
	at java.base/sun.security.pkcs.SignerInfo.verify(SignerInfo.java:324)
	at java.base/sun.security.pkcs.PKCS7.verify(PKCS7.java:578)
	at java.base/sun.security.pkcs.PKCS7.verify(PKCS7.java:595)
	at java.base/sun.security.util.SignatureFileVerifier.processImpl(SignatureFileVerifier.java:293)
	at java.base/sun.security.util.SignatureFileVerifier.process(SignatureFileVerifier.java:269)
	at java.base/java.util.jar.JarVerifier.processEntry(JarVerifier.java:316)
	at java.base/java.util.jar.JarVerifier.update(JarVerifier.java:230)
	at java.base/java.util.jar.JarFile.initializeVerifier(JarFile.java:749)
	at java.base/java.util.jar.JarFile.ensureInitialization(JarFile.java:1016)
	at java.base/java.util.jar.JavaUtilJarAccessImpl.ensureInitialization(JavaUtilJarAccessImpl.java:72)
	at java.base/jdk.internal.loader.URLClassPath$JarLoader$2.getManifest(URLClassPath.java:873)
	at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(BuiltinClassLoader.java:811)
	at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:723)
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:646)
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:604)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:168)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at io.crate.types.DataTypeXContentExtension.getXContentWriters(DataTypeXContentExtension.java:40)
	at org.elasticsearch.common.xcontent.XContentBuilder.<clinit>(XContentBuilder.java:119)
	at org.elasticsearch.common.settings.Setting.arrayToParsableString(Setting.java:1308)
	at org.elasticsearch.common.settings.Setting$ListSetting.lambda$new$0(Setting.java:1333)
	at org.elasticsearch.common.settings.Setting.<init>(Setting.java:151)
	at org.elasticsearch.common.settings.Setting$ListSetting.<init>(Setting.java:1330)
	at org.elasticsearch.common.settings.Setting.listSetting(Setting.java:1282)
	at org.elasticsearch.common.settings.Setting.listSetting(Setting.java:1250)
	at org.elasticsearch.env.Environment.<clinit>(Environment.java:55)
	at org.elasticsearch.node.InternalSettingsPreparer.prepareEnvironment(InternalSettingsPreparer.java:66)
	at io.crate.bootstrap.CrateDB.createEnv(CrateDB.java:112)
	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:82)
	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124)
	at org.elasticsearch.cli.Command.main(Command.java:90)
	at io.crate.bootstrap.CrateDB.main(CrateDB.java:91)
	at io.crate.bootstrap.CrateDB.main(CrateDB.java:84)
Caused by: java.lang.NullPointerException: Cannot invoke "java.io.InputStream.available()" because "is" is null
	at java.base/jdk.internal.icu.impl.ICUBinary.getRequiredData(ICUBinary.java:94)
	at java.base/jdk.internal.icu.impl.NormalizerImpl.load(NormalizerImpl.java:431)
	at java.base/jdk.internal.icu.impl.Norm2AllModes$Norm2AllModesSingleton.<init>(Norm2AllModes.java:274)
	at java.base/jdk.internal.icu.impl.Norm2AllModes$NFKCSingleton.<clinit>(Norm2AllModes.java:290)
	at java.base/jdk.internal.icu.impl.Norm2AllModes.getNFKCInstance(Norm2AllModes.java:263)
	at java.base/jdk.internal.icu.text.Normalizer2.getNFKDInstance(Normalizer2.java:123)
	at java.base/jdk.internal.icu.text.NormalizerBase$NFKDModeImpl.<clinit>(NormalizerBase.java:181)
	... 48 more

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:run'.
> Process 'command 'C:\Users\Baur\.gradle\caches\transforms-3\00cf94b11d7d07694077c263b7fcc56e\transformed\windows-15.0.2-x64.zip/bin/java'' finished with non-zero exit value 1
```
</details>

I've been investigating and trying to fix both commands failure  - if I manage to make itest related scripts cross-platform I will create another PR for that.

It would be nice if we can trigger remaining ` jenkins/pr_tests` check to see how my manual additions to the `postgres.rst `will behave in a proper environment.

amotl commented:
Hi @BaurzhanSakhariev,

thanks for your efforts. Regarding blackbox testing, I believe it is using [`cr8`](https://github.com/mfussenegger/cr8) under the hood. In order to bring that to Windows (see https://github.com/mfussenegger/cr8/issues/327), I already started https://github.com/mfussenegger/cr8/pull/337 but haven't been able to complete this yet.

This is just to give you a heads-up which kinds of patches would have to be applied as a minimum to make it work on Windows. The state of the onion is that, by using those patches, it should already work reasonably but currently fails on test layer teardown on GHA/CI due to some nasty deadlocking issue where I haven't been able to investigate further.

So, while it will probably be a good idea to continue day-to-day development of CrateDB on Linux or macOS, I am happy to hear you also know your way around Windows and hope that you will be available to improve Windows support for all the test harnesses we are running as well as Windows support in general. See, for example, #10516 and #6349.

With kind regards,
Andreas.

seut commented:
retest this please

BaurzhanSakhariev commented:
Hi @amotl,

Thanks for the detailed comment! You provided a lot of useful information, gonna digest it)

BaurzhanSakhariev commented:
retest this please

BaurzhanSakhariev commented:
retest this please

BaurzhanSakhariev commented:
I managed to run locally, and checked result of performing `SELECT ` in `crash ` with my manual updates - I forgot to update number of rows (`SELECT 43 rows in set`). Can someone please trigger pr_tests check? I left a comment `retest this please` but check was not triggered, probably because I don't have Jenkins account.

Regarding my issues with run and itest:

run: Previously I had only Java 11 on my machine, `gradlew app:run `worked once I installed Java 15. Not sure how it helped, bundled Java 15 was used before.

itest - yet to fix on my machine, I keep investigating.
so far I found 2 differences for Windows:
[bin->Scripts](https://github.com/pypa/virtualenv/commit/993ba1316a83b760370f5a3872b3f5ef4dd904c1) 
and using `python `instead of `python3.7` or python3 in `boostrap.sh` - Python installer creates only `python `binary and I tried [simply renaming](https://stackoverflow.com/a/55166698/2104560) it but switched back to changing script to use `python `because [renaming was not good idea](https://github.com/msys2/MINGW-packages/issues/5001#issuecomment-478027026)

seut commented:
retest this please

BaurzhanSakhariev commented:
retest this please

BaurzhanSakhariev commented:
retest this please

seut commented:
retest this please

seut reviewed (approved):
👍 Looks good, thanks a again for this nice contribution.

BaurzhanSakhariev commented:
retest this please

BaurzhanSakhariev commented:
retest this please

seut commented on server/src/main/java/io/crate/expression/scalar/CurrentDateFunction.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Override
+    @SafeVarargs
+    public final LocalDate evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>... args) {
```
We've just saw that this seems odd as the returned data type here must be a CrateDB data type and must align with the signatures. So I should be `Date` instead. Isn't `current_date` currently failing because of that when used in full integration test or manual test by e.g. using `crash`?

BaurzhanSakhariev replied:
Good catch, thanks.
<details>

<summary> current_date works locally (screenshot under the spoiler) </summary>

![Capture](https://user-images.githubusercontent.com/7446940/109770117-6ff0ec80-7bfb-11eb-943a-08a31ed613b0.PNG)

</details>

However, `current_date` calls `DateType.implicitCast` which tries to cast `LocalDate` to `Long` causing Exception in `Scalar.normalizeSymbol`. It works because exception case is covered but that casting issue must be avoided indeed. 

I have tried locally 3 approaches:

1. Keep current implementation but add "instance of LocalDate' check to avoid casting exception
2. 
pros 

- no need to create wrapper class
- simple Long type is used as internal type 

cons

- no control over output format - basically it's LocalDate's default toString which uses `ISO-8601`
- as you mentioned, signature mismatch looks odd

2. 
Use wrapper class (not LocalDate but own lightweight class, just long millis holder) - idea similar to TimeTZ class
pros
- control over format - can specify any format in `toString` and later improve `curdate()` to accept date format.
- signature is still mismatching with `module.register` but I guess in this in case it's better - as evaluate return type matches with `extends DataType<WrapperClass>`

 cons
 
 - <details>
 <summary>curdate() returns 2 rows as a result: `toString() + timestamp(?)` - haven't figured out how to avoid this.</summary>
 
 
![CaptureUtcDate](https://user-images.githubusercontent.com/7446940/109825565-3be8ec00-7c3a-11eb-9f3b-4d4e0a58aa31.PNG)

 </details>

3. Do not create wrapper class but just return DateType in evaluate. Add long millis as a fiield and add toString with needed format.

pros 
- signature matches and no need to create wrapper class
cons  
- not consistent with other functions - I haven't seen any function returning not internal type but type itself, even current_time uses not TimeTZType but TimeTZ.


Let me know which option is better

BaurzhanSakhariev replied:
I implemented second option from my comment above. 

<details>

<summary>
I tested it locally, it returns date and also timestamp - 
</summary>

![Capture_all_3_with_second_row](https://user-images.githubusercontent.com/7446940/110487973-5fa2ab00-80ee-11eb-86da-bf5e73083daa.PNG)

</details

As showed in the screenshot, all time/date related functions return 2 rows  - I couldn't get rid of timestamp but as `current_time` and `current_timestamp` also return 2 rows I assumed that it's expected behavior.

seut replied:
@BaurzhanSakhariev 

Sorry for the late reply and it also seems that I was also confused by the `current_time` implementation and pushed you maybe in the wrong direction.

`current_time` will respond an `array(long)` where the 1st item is the current time since epoch and the 2nd the timezone offset. That is needed by clients to render the current time correctly.

But for `current_date` no timezone information is needed and exists.

So we can just return a `Long` on the evaluate method and no wrapper or `rest` specific json formatting is needed. (which we did for `TimeTZ`, see https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/types/DataTypeXContentExtension.java#L67).

seut replied:
Making it human readable is up to the client, so `crash` should do that which it currently doesn't (we should create an issue/fix there). Otherwise the client would have to parse a human readable format even if it would never be presented to humans.

BaurzhanSakhariev replied:
Ah, I see, I misunderstood the task. 
I saw select example in the [issue ](https://github.com/crate/crate/issues/9651) and had a wrong impression that function itself should deliver ready-to-present result. That clarification makes things simpler, thanks!

seut replied:
> I saw select example in the issue and had a wrong impression that function itself should deliver ready-to-present result.

Ah I see. Sorry for all this confusion!

seut commented:
@BaurzhanSakhariev Already approved it but figured out that there is still something todo. I've left a comment related to the return type of the scalar.
The other last thing missing: we should document the new data type here https://github.com/crate/crate/blob/master/docs/interfaces/http.rst#column-types.

BaurzhanSakhariev commented:
> The other last thing missing: we should document the new data type here https://github.com/crate/crate/blob/master/docs/interfaces/http.rst#column-types.

Thanks for pointing it out, spot missing date type section in `ddl/data-types.rst` while updating `http.rst` and found an issue with casts. Updated `DataTypes.SAFE_CONVERSIONS` and `DataTypes.NUMBER_CONVERSIONS` to address that.

seut commented on docs/general/ddl/data-types.rst:
```diff
@@ -561,6 +565,37 @@ Time zone syntax as defined by `ISO 8601 time zone designators`_.
 
+.. _date-data-type:
+
+date
+-------------------
```
```suggestion
``date``
---------
```

seut commented on docs/general/ddl/data-types.rst:
```diff
@@ -561,6 +565,37 @@ Time zone syntax as defined by `ISO 8601 time zone designators`_.
+date
+-------------------
+
+A string literal that contain a date will be
+converted to UTC date.
```
```suggestion
.. NOTE::

    The storage of the ``date`` data type is not supported. Therefore,
    it is not possible to create tables with ``date`` fields.

The ``date`` data type consist of year, month, and day.
A constant value of type ``date`` can be expressed by explicit casts 
of string or numeric literals.

The string format for dates is `YYYY-MM-DD`. For example `2021-03-09`.
```

seut commented on docs/general/ddl/data-types.rst:
```diff
@@ -561,6 +565,37 @@ Time zone syntax as defined by `ISO 8601 time zone designators`_.
+    SELECT 1 row in set (... sec)
+
+
+Date also accepts a ``bigint`` representing UTC milliseconds since
+the epoch.
+
+::
+
+    cr> select 1615248000000::date as cd;
+   +---------------+
+    |            cd |
+    +---------------+
+    | 1615248000000 |
+    +---------------+
+    SELECT 1 row in set (... sec)
```
I'd remove this example if you take my suggestion above which already mention the support of numeric->date casts.

BaurzhanSakhariev replied:
Done. I removed string casting example `select '2021-03-09'::date` as well as it's also mentioned in suggestion.

seut replied:
I think keeping the string cast example would still be valuable (also this may be the common use case).

BaurzhanSakhariev replied:
done

seut commented on docs/general/ddl/data-types.rst:
```diff
@@ -668,6 +703,7 @@ For example::
     +-------------------------------+
     | result                        |
     +-------------------------------+
+    +-------------------------------+
```
Mistake? ;)

BaurzhanSakhariev replied:
yes, conflicts resolving mistake, thanks for spotting it.

seut commented on server/src/main/java/io/crate/expression/scalar/CurrentDateFunction.java:
```diff
@@ -0,0 +1,71 @@
+            Signature.scalar(
+                NAME,
+                DataTypes.DATE.getTypeSignature()
+            ),
```
```suggestion
            ).withFeatures(NO_FEATURES),
```

This will ensure that the function won't be normalized and as a side-effect, will preserve the function name as the column name if no alias is used.

seut reviewed (commented):
Thanks for your changes. I've added some comments mostly related to documentation.

BaurzhanSakhariev commented:
retest this please

seut commented:
retest this please

mfussenegger commented on docs/general/ddl/data-types.rst:
```diff
@@ -561,6 +565,31 @@ Time zone syntax as defined by `ISO 8601 time zone designators`_.
+    The storage of the ``date`` data type is not supported. Therefore,
+    it is not possible to create tables with ``date`` fields.
+
+The ``date`` data type consist of year, month, and day.
+A constant value of type ``date`` can be expressed by explicit casts 
+of string or numeric literals.
+
+The string format for dates is `YYYY-MM-DD`. For example `2021-03-09`.
```
```suggestion
The ``date`` data type can be used to represent values with a year, month and a day.

To construct values of the type ``date`` you can cast a string literal or a numeric literal to ``date``.
If a numeric value is used, it must contain the number of milliseconds since ``1970-01-01T00:00:00Z``.

The string format for dates is `YYYY-MM-DD`. For example `2021-03-09`.
```

mfussenegger commented on server/src/main/java/io/crate/expression/scalar/CurrentDateFunction.java:
```diff
@@ -0,0 +1,70 @@
+    @Override
+    @SafeVarargs
+    public final Long evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>... args) {
+        return txnCtx.currentInstant().toEpochMilli();
```
I think the value needs to be truncated to the date, no?

```
cr> select date_trunc('day', current_timestamp), current_date;
+--------------------------------------+---------------+
| date_trunc('day', CURRENT_TIMESTAMP) |     curdate() |
+--------------------------------------+---------------+
|                        1615766400000 | 1615800127721 |
+--------------------------------------+---------------+
```

BaurzhanSakhariev replied:
Thanks for spotting this, forgot to update logic when changed LocalDate to Long. Added a test for that case.

mfussenegger commented on server/src/main/java/io/crate/protocols/postgres/types/DateType.java:
```diff
@@ -0,0 +1,79 @@
+    Object decodeUTF8Text(byte[] bytes) {
+        String s = new String(bytes, StandardCharsets.UTF_8);
+
+        LocalDate dt = LocalDate.parse(s, ISO_FORMATTER);
+        return dt.atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli();
```
PostgreSQL supports more text formats as input:

https://www.postgresql.org/docs/current/datatype-datetime.html

I think we either have to extend this, or make it explicit in the documentation that the text format support is not compatible with all PostgreSQL clients.

BaurzhanSakhariev replied:
I updated `data-types.rst` and added TODO to `DateType`

mfussenegger reviewed (commented):
Left some more coments. 
Overall seems like it's almost ready :+1:

BaurzhanSakhariev commented:
retest this please

mfussenegger commented:
ok to test

mfussenegger reviewed (approved):
Nice work :+1:

seut reviewed (approved):
👍 Great!
